### ISSUE TITLE
Is there a way to get insights into traces?

### ISSUE THREAD

**Issue by lfdversluis** [ben-manes/caffeine#331]
Is there any graphing tooling in the simulator to visualize the input traces? As far as I understand, traces are a list of object keys (longs) that are being fetched. I saw the discussion on the scarab traces and some graphs were shown there, is there tooling available that generate these and more?

In no particular order: I am curious about the number of keys being requested over time, a distribution of how often the same key is being requested in the trace, and the graphs shown there with ideal hit rates per cache size and whatnot.

**Comment by ben-manes** [comment]
I typically use Google Spreadsheets to visualize, which has been the most straightforward. The current output can be dumped in an ascii table or csv, making it easy to drop into a chart. Often I am doing an ad hoc, one time exploration so it hasn't been worth the effort to automate much more.

Other than the summary hit rates, I sampled them for the adaptive chart to see how well the policy reacted over time. That was a one-off when exploring that topic.

A quick way to analyze the traces themselves is to plug in another policy that dumps those metrics. I've done this ad hoc when translating to another simulator's format to run compare to that author's policy, rather than implement it directly (e.g. LHD).

In short, there is no direct tooling but it should be easy to write in something to collect metrics as desired. If you consider it useful longer term, we can generalize it a bit and make it a feature. For visualizations, it probably remains simplest to use a spreadsheet instead of trying to generate them directly (or equivalent, like csv => gnuplot).

**Comment by ben-manes** [comment]
mimircache.info has a heatmap visualization which you might find interesting

**Comment by lfdversluis** [comment]
I am trying to cook up some graphs that could be interesting to add as a module to the simulator using JFreeChart, but that library is not extensive as popular python libraries, unfortunately. I will check if I can make something PR worthy.

**Comment by ben-manes** [comment]
I suspect you'll have a much easier and enjoyable time if you generate data and post-process it. As you said, Java just doesn't have great libraries for this and most reporters don't even try anymore. For example the gradle-jmh-report plugin creates an html report from the JMH benchmark results (json). It uses [recharts](http://recharts.org/en-US/) to visualize, as JS has numerous charting libraries to choose from. If you are concerned with local dependencies (like Python) then instead you can wrap it up in a tiny webapp, invoke it via a Java http client, and leverage free hosting to run indefinitely. Lots of choices if doing it in pure Java is too painful.
